pinot-common/src/main/java/com/linkedin/pinot/common/data/FieldSpec.java

```diff
@@ -15,152 +15,209 @@
 import org.json.JSONException;
 import org.json.JSONObject;
 
 
-// Json annotation required for abstract classes.
-//@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "type")
 public abstract class FieldSpec {
   private static final String DEFAULT_DIM_NULL_VALUE_OF_STRING = "null";
-  private static final Integer DEFAULT_DIM_NULL_VALUE_OF_INT = Integer.valueOf(Integer.MIN_VALUE);
-  private static final Long DEFAULT_DIM_NULL_VALUE_OF_LONG = Long.valueOf(Long.MIN_VALUE);
-  private static final Float DEFAULT_DIM_NULL_VALUE_OF_FLOAT = Float.valueOf(Float.NEGATIVE_INFINITY);
-  private static final Double DEFAULT_DIM_NULL_VALUE_OF_DOUBLE = Double.valueOf(Double.NEGATIVE_INFINITY);
-
-  private static final Integer DEFAULT_METRIC_NULL_VALUE_OF_INT = Integer.valueOf(0);
-  private static final Long DEFAULT_METRIC_NULL_VALUE_OF_LONG = Long.valueOf(0);
-  private static final Float DEFAULT_METRIC_NULL_VALUE_OF_FLOAT = Float.valueOf(0);
-  private static final Double DEFAULT_METRIC_NULL_VALUE_OF_DOUBLE = Double.valueOf(0);
-
-  String name;
-  FieldType fieldType;
-  DataType dataType;
-  boolean isSingleValueField = true;
-  String delimiter = ",";
-  Object defaultNullValue;
+  private static final Integer DEFAULT_DIM_NULL_VALUE_OF_INT = Integer.MIN_VALUE;
+  private static final Long DEFAULT_DIM_NULL_VALUE_OF_LONG = Long.MIN_VALUE;
+  private static final Float DEFAULT_DIM_NULL_VALUE_OF_FLOAT = Float.NEGATIVE_INFINITY;
+  private static final Double DEFAULT_DIM_NULL_VALUE_OF_DOUBLE = Double.NEGATIVE_INFINITY;
+
+  private static final Integer DEFAULT_METRIC_NULL_VALUE_OF_INT = 0;
+  private static final Long DEFAULT_METRIC_NULL_VALUE_OF_LONG = 0L;
+  private static final Float DEFAULT_METRIC_NULL_VALUE_OF_FLOAT = 0.0F;
+  private static final Double DEFAULT_METRIC_NULL_VALUE_OF_DOUBLE = 0.0D;
+
+  private String _name;
+  private FieldType _fieldType;
+  private DataType _dataType;
+  private boolean _isSingleValueField = true;
+  private String _delimiter = ",";
+  private Object _defaultNullValue;
+
+  // Fields to help make sure we first set data type then default null value.
+  @JsonIgnore
+  protected boolean _isDataTypeSet = false;
+  @JsonIgnore
+  protected boolean _isDefaultNullValueSet = false;
 
   public FieldSpec() {
-
   }
 
-  public FieldSpec(String name, FieldType fType, DataType dType, boolean singleValue, String delimeter) {
-    this.name = name;
-    this.fieldType = fType;
-    this.dataType = dType;
-    this.isSingleValueField = singleValue;
-    this.delimiter = delimeter;
+  public FieldSpec(String name, FieldType fieldType, DataType dataType, boolean isSingleValueField, String delimiter) {
+    _name = name;
+    _fieldType = fieldType;
+    _dataType = dataType;
+    _isSingleValueField = isSingleValueField;
+    _delimiter = delimiter;
+    _isDataTypeSet = true;
   }
 
-  public FieldSpec(String name, FieldType fType, DataType dType, boolean singleValue, String delimeter,
```

Comment:
I removed it because no one use this constructor. But you are right, we should keep such a constructor for convenience. Will fix it.